common/junit-platform-native/src/main/java/org/graalvm/junit/platform/JUnitPlatformFeature.java

```diff
@@ -85,37 +84,49 @@ public void duringSetup(DuringSetupAccess access) {
     public void beforeAnalysis(BeforeAnalysisAccess access) {
         RuntimeClassInitialization.initializeAtBuildTime(NativeImageJUnitLauncher.class);
 
+        List<Path> classpathRoots = access.getApplicationClassPath();
+        List<? extends DiscoverySelector> selectors = getSelectors(classpathRoots);
+
         Launcher launcher = LauncherFactory.create();
-        TestPlan testplan = registerTestPlan(launcher, access.getApplicationClassPath());
+        TestPlan testplan = discoverTestsAndRegisterTestClassesForReflection(launcher, selectors);
         ImageSingletons.add(NativeImageJUnitLauncher.class, new NativeImageJUnitLauncher(launcher, testplan));
     }
 
-    private List<? extends DiscoverySelector> getSelectors(List<Path> classpath) {
-        InputStream listenerStream;
+    private List<? extends DiscoverySelector> getSelectors(List<Path> classpathRoots) {
         try {
-            listenerStream = new FileInputStream(UniqueIdTrackingTestExecutionListener.FILE_NAME);
-        } catch (FileNotFoundException e) {
-            listenerStream = getClass().getClassLoader().getResourceAsStream(UniqueIdTrackingTestExecutionListener.FILE_NAME);
-        }
-        try (BufferedReader br = new BufferedReader(new InputStreamReader(listenerStream))) {
-            List<String> classes = br.lines().collect(Collectors.toList());
-            System.out.println("[junit-platform-native] Running in 'test listener' mode.");
-            return classes.stream().map(DiscoverySelectors::selectUniqueId).collect(Collectors.toList());
-        } catch (IOException | NullPointerException e) {
-            System.out.println("[junit-platform-native] Running in 'test discovery' mode. Note that this is a fallback mode.");
+            Path outputDir = new UniqueIdTrackingListener().getOutputDir();
```

Comment:
That's only true in this variant of the listener. The official listener actually uses `ConfigurationParameters` available in the `TestPlan` beginning with JUnit Platform 1.8.

https://github.com/junit-team/junit5/blob/9a1db7309f908f474b6042c9eb68c1e596e58453/junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/UniqueIdTrackingListener.java#L197

Thus, we will require a different approach once we switch to the official listener.

My goal here was to get something in place that works now.

Once we switch to the official listener, I think the Gradle and Maven plugins should always specify an explicit output directory (instead of the way the current implementation of the Maven plugin relies on the default output directory selection based on convention). Then we would no longer need to rely on the behavior in `getOutputDir()`.

Does that make sense?

Reply:
Ah, in that case, it makes sense to leave it as is - I agree, the build tool should explicitly specify this directory